src/HttpClient/HttpClient.php

```diff
@@ -0,0 +1,89 @@
+declare(strict_types=1);
+
+namespace Sentry\HttpClient;
+
+use Sentry\Options;
+use Sentry\Util\Http;
+
+/**
+ * @internal
+ */
+class HttpClient implements HttpClientInterface
+{
+    /**
+     * @var string The Sentry SDK identifier
+     */
+    protected $sdkIdentifier;
+
+    /**
+     * @var string The Sentry SDK identifier
+     */
+    protected $sdkVersion;
+
+    public function __construct(string $sdkIdentifier, string $sdkVersion)
+    {
+        $this->sdkIdentifier = $sdkIdentifier;
+        $this->sdkVersion = $sdkVersion;
+    }
+
+    public function sendRequest(string $requestData, Options $options): Response
+    {
+        $dsn = $options->getDsn();
+        if (null === $dsn) {
+            throw new \RuntimeException('The DSN option must be set to use the HttpClient.');
+        }
+
+        $curlHandle = curl_init();
+        curl_setopt($curlHandle, \CURLOPT_URL, $dsn->getEnvelopeApiEndpointUrl());
+        curl_setopt($curlHandle, \CURLOPT_HTTPHEADER, Http::getRequestHeaders($dsn, $this->sdkIdentifier, $this->sdkVersion));
+        curl_setopt($curlHandle, \CURLOPT_USERAGENT, $this->sdkIdentifier . '/' . $this->sdkVersion);
+        curl_setopt($curlHandle, \CURLOPT_TIMEOUT, $options->getHttpTimeout());
+        curl_setopt($curlHandle, \CURLOPT_CONNECTTIMEOUT, $options->getHttpConnectTimeout());
+        curl_setopt($curlHandle, \CURLOPT_ENCODING, '');
+        curl_setopt($curlHandle, \CURLOPT_POST, true);
+        curl_setopt($curlHandle, \CURLOPT_POSTFIELDS, $requestData);
+        curl_setopt($curlHandle, \CURLOPT_RETURNTRANSFER, true);
+        curl_setopt($curlHandle, \CURLOPT_HEADER, true);
+        curl_setopt($curlHandle, \CURLOPT_HTTP_VERSION, \CURL_HTTP_VERSION_1_1);
+
+        $httpSslVerifyPeer = $options->getHttpSslVerifyPeer();
+        if ($httpSslVerifyPeer) {
+            curl_setopt($curlHandle, \CURLOPT_SSL_VERIFYPEER, true);
+        }
+
+        $httpProxy = $options->getHttpProxy();
+        if (null !== $httpProxy) {
+            curl_setopt($curlHandle, \CURLOPT_PROXY, $httpProxy);
+        }
+
+        $httpProxyAuthentication = $options->getHttpProxyAuthentication();
+        if (null !== $httpProxyAuthentication) {
+            curl_setopt($curlHandle, \CURLOPT_PROXYUSERPWD, $httpProxyAuthentication);
+        }
+
+        /**
+         * @TODO(michi) add request compression (gzip/brotli) depending on availiable extensions
+         */
+        $body = curl_exec($curlHandle);
+
+        if (false === $body) {
+            $errorCode = curl_errno($curlHandle);
+            $error = curl_error($curlHandle);
+            curl_close($curlHandle);
+
+            $message = 'cURL Error (' . $errorCode . ') ' . $error;
+
+            return new Response(0, [], $message);
+        }
+
+        $statusCode = curl_getinfo($curlHandle, \CURLINFO_HTTP_CODE);
+        $headerSize = curl_getinfo($curlHandle, \CURLINFO_HEADER_SIZE);
+        $headers = Http::getResponseHeaders($headerSize, (string) $body);
```

Comment:
~~I think we should use `\CURLOPT_HEADERFUNCTION` instead. This gives us a callback for each headerline already normalized to lowercase names: https://curl.se/libcurl/c/CURLOPT_HEADERFUNCTION.html. We would also want to set `CURLOPT_SUPPRESS_CONNECT_HEADERS` for the proxy requests.~~

Realized this is a PHP 7.3+ option, so not possible for now but should make header parseing more reliable.

Reply:
Why? Do we still want to support PHP 7.2 and below even in the next major?

Reply:
I made a mistake `CURLOPT_HEADERFUNCTION` is available way longer, not starting in PHP 7.3, I've changed my comment to reflect that 🤦

Reply:
I like this a lot, thanks Alex! Have a look at https://github.com/getsentry/sentry-php/pull/1589/commits/c9b3a2da444767b2793c5994a228e92c1675d4c1